inc/scroom/layeroperations.hh

```diff
@@ -142,3 +142,18 @@ public:
   virtual void reduce(Tile::Ptr target, const ConstTile::Ptr source, int x, int y);
 };
 
+class OperationsCMYK : public CommonOperations
```

Comment:
- I recommend you make separate classes for the 1, 2, 4 and 8 bps cases. They share hardly any code
- I also recommend you choose your types such that you need far fewer casts. The C++ language designers made casts look ugly (compared to C) to encourage you to avoid them were possible
- Take a look at the [PixelIterator](https://kees-jan.github.io/scroom/0.10/classScroom_1_1Bitmap_1_1PixelIterator.html) class. When used as a SampleIterator, it should save you a fair amount of code
- When reducing a 1, 2 or 4 bps layer, store the result in a 8bps layer. That'll look better and save you from some rounding errors

Reply:
- I noticed that the RGB Operation classes are all split based on the number of bits per sample. I think they have quite a bit of code duplication, with just a few constants being different. With the "Don't Repeat Yourself" creed in mind, I figured it'd be better to just have one class that handles all different bitdepths. At that time, I noticed the PixelIterator class, and I expected it to be able to abstract those differences away.
- I agree that type casts shouldn't be littered throughout the code. Sometimes however, adding casts avoids warning and overflows. Sometimes, you simply need data types smaller than the default `int`. I think this is one of those cases. Please let me know if you have a good way to save specific casts.
- Unfortunately, I can't get the PixelIterator to work for CMYK. Sadly, not all existing Operations (sub)classes use the PixelIterator and there is not really documentation on how to use it. Prompted by this comment, I tried it again, but to no avail. I could not find any reference to "SampleIterator" in the code / documentation. I guess you meant that I should create a PixelIterator whose "Base" type is a type the individual samples precisely fit in (which would only work for 8bps, unfortunately). However, the number of pixels per byte is now less than 1, which means that it is rounded to 0 because of integer division. This causes various problems, most importantly a division by 0 in the constructor.
- I agree that it helps with rounding (and thus makes it look better). However, I could not figure out how to make sure the Tile created by the reduce function has a different bitdepth from the source Tile.

Reply:
_Regarding splitting this class_
As it is, these methods are fairly large. Near as I can tell, there would be hardly any shared code if you split them. If you decide to go the other way and abstract the differences away, that’s also fine. It is just that, as it is, it is not very maintainable, and I can’t imagine it would score well on the metrics laid out by the university.

_Regarding PixelIterator_
If there’s anything in the codebase that you don’t understand or can’t get to work, feel free to ask, I’m happy to help. When you do ask questions, please make them as specific as possible, detailing what you tried, what you expected to happen, and what actually happen. “Unfortunately there is no documentation” (technically not even a question :wink:) is a bit to broad to tackle in limited time

I’ve noticed that there aren’t yet any tests for the PixelIterator class. If you were to contribute some tests that are failing, even though in your opinion they shouldn’t be, I’d be happy to look at them.

_Regarding casts_
The more different integer types you use, the more you will need to cast them. My recommendation would be to stick to (unsigned) integer, and only cast them when storing in the target memory location

The default should be that there are no casts at all. That means that it is not my task to suggest specific improvements, but yours to defend any casts that are currently in the code.

_Regarding rounding and layers with different bpp values_
Just add a second LayerOperations object with a different bpp value to your LayerSpec. Then the subsequent layers will have that bpp value. See the 1bpp grayscale implementation for an example

Reply:
_Regarding splitting `OperationsCMYK`_
I've split the class up in 4 classes in commit 27e92a12244c6fb6408e07e0f78329e48c153ada. I think that resolves that bullet point.

_Regarding PixelIterator_
I think making separate classes for each of the different bitdepths removes the need for a `PixelIterator`. I'm still unsure what you meant by `SampleIterator`.

_Regarding casts_
By moving the various bitdepths to different classes, most casts have disappeared. As such, I also think that this bullet point is resolved.

Reply:
I think this code still has unneeded casts, and could still benefit from a PixelIterator.

But I'm giving up, for now :wink: